rational/src/lib.rs

```diff
@@ -668,6 +668,176 @@ impl RatioErrorKind {
+    fn from_i64(n: i64) -> Option<Self> {
+        Some(Ratio::from_integer(n.into()))
+    }
+
+    fn from_u64(n: u64) -> Option<Self> {
+        Some(Ratio::from_integer(n.into()))
+    }
+
+    fn from_f32(n: f32) -> Option<Self> {
+        Ratio::from_float(n)
+    }
+
+    fn from_f64(n: f64) -> Option<Self> {
+        Ratio::from_float(n)
+    }
+}
+
+macro_rules! from_primitive_integer {
+    ($typ:ty, $approx:ident) => {
+        impl FromPrimitive for Ratio<$typ> {
+            fn from_i64(n: i64) -> Option<Self> {
+                <$typ as FromPrimitive>::from_i64(n).map(Ratio::from_integer)
+            }
+
+            fn from_u64(n: u64) -> Option<Self> {
+                <$typ as FromPrimitive>::from_u64(n).map(Ratio::from_integer)
+            }
+
+            fn from_f32(n: f32) -> Option<Self> {
+                $approx(n, 10e-20, 30)
+            }
+
+            fn from_f64(n: f64) -> Option<Self> {
+                $approx(n, 10e-20, 30)
+            }
+        }
+    }
+}
+
+from_primitive_integer!(i8, approximate_float);
+from_primitive_integer!(i16, approximate_float);
+from_primitive_integer!(i32, approximate_float);
+from_primitive_integer!(i64, approximate_float);
+from_primitive_integer!(isize, approximate_float);
+
+from_primitive_integer!(u8, approximate_float_unsigned);
+from_primitive_integer!(u16, approximate_float_unsigned);
+from_primitive_integer!(u32, approximate_float_unsigned);
+from_primitive_integer!(u64, approximate_float_unsigned);
+from_primitive_integer!(usize, approximate_float_unsigned);
+
+impl<T: Integer + Signed + Bounded + NumCast + Clone> Ratio<T> {
+    pub fn approximate_float<F: Float + NumCast>(f: F) -> Option<Ratio<T>> {
+        // 1/10e-20 < 1/2**32 which seems like a good default, and 30 seems
+        // to work well. Might want to choose something based on the types in the future, e.g.
+        // T::max().recip() and T::bits() or something similar.
+        let epsilon = <F as NumCast>::from(10e-20).expect("Can't convert 10e-20");
+        approximate_float(f, epsilon, 30)
+    }
+}
+
+fn approximate_float<T, F>(val: F, max_error: F, max_iterations: usize) -> Option<Ratio<T>>
+    where T: Integer + Signed + Bounded + NumCast + Clone,
+          F: Float + NumCast
+{
+    let negative = val.is_sign_negative();
+    let abs_val = val.abs();
+
+    let r = approximate_float_unsigned(abs_val, max_error, max_iterations);
+
+    // Make negative again if needed
+    if negative {
+        r.map(|r| r.neg())
+    } else {
+        r
+    }
+}
+
+// No Unsigned constraint because this also works on positive integers and is called
+// like that, see above
+fn approximate_float_unsigned<T, F>(val: F, max_error: F, max_iterations: usize) -> Option<Ratio<T>>
+    where T: Integer + Bounded + NumCast + Clone,
+          F: Float + NumCast
+{
+    // Continued fractions algorithm
+    // http://mathforum.org/dr.math/faq/faq.fractions.html#decfrac
+
+    assert!(val >= F::zero());
```

Comment:
Not sure I understand what you mean here. The signed ones all go into approximate_float(), which calls approximate_float_unsigned() for the absolute value and then negates the result if the original value was negative. approximate_float_unsigned() asserts that the input value is >= 0, i.e. "unsigned".